More complete meeting. Fixes #804.

rgrig opened the pull request:
This PR does a more complete meet.

The current/proposed code has several disadvantages compared to the old one: it is longer, it is more complicated, it is still incomplete. However, the current code has an important advantage: it is incomplete only because of timeout. In particular, it is complete for situations in which the number specs is small.

According to a test on Tomcat, the overall speed is not affected, and the same issues are reported (plus an extra NULL_DEREFERENCE).

jeremydubreil commented on infer/src/backend/dom.ml:
```diff
@@ -2173,55 +2173,78 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
   in
   res
 
+let maximal_combine (timeout : int) (plus : 'a -> 'a -> 'a option) (xs : 'a list)  : 'a list =
```
@rgrig: Thanks for the pull request.

Can you please add comments in the code to explain what the code is doing. For maintainability, we can only include code that people can understand and modify quickly.

rgrig replied:
@jeremydubreil: I started adding some comments and noticed this is less complete than I thought it is. So, thanks for asking :) I should be able to update this PR before Monday.

rgrig commented:
Updated.

mbouaziz commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    if !maxclique_count >= timeout then raise Timeout;
+    let xs, ng = next_maxclique g in (* xs is a pairwise consistent set *)
+    let zs = List.map ~f:by_idx xs in
+    (try result :=  (List.fold ~init ~f:plus_exc zs) :: !result
```
I am wondering if a cache of prefixes would make sense

rgrig replied:
seems to: cache hit quite high
(also, could be used for the vertex cover, if needed later)

mbouaziz commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    incr maxclique_count;
+    if !maxclique_count >= timeout then raise Timeout;
+    let xs, ng = next_maxclique g in (* xs is a pairwise consistent set *)
+    let zs = List.map ~f:by_idx xs in
```
You can have a `plus_by_idx` to avoid creating this temporary list

rgrig replied:
done

mbouaziz commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+  let plus_exc x y = match plus x y with
+    | None -> raise Undefined
+    | Some z -> z in
+  let pairs = (* first, build graph with pairwise consistency *)
```
Please have `pairs` defined in `edge`, like you did with `by_idx`

rgrig replied:
This became the `cache` now, and it is inside `plus_by_idx`.

rgrig commented:
I'll not have time to look at this for a few days.

Briefly:
 * on Tomcat, there isn't an observable performance penalty
 * the examples where precision increases are similar to #804 

For code, the plan with `pairs` was to eventually use it also as a cache for prefixes, but then it seemed like such a cache is not needed. I'll try it, though.

Maximal cliques are inconsistent extremely rarely (<1%), which is why I was comfortable to throwing them away. But, yes, it would be better to still keep something. (Either way is sound, though.) I'll try what you say (keep longest consistent prefix).

rgrig commented:
I added a test, added a cache for prefixes, and refactored to avoid some effects. Also rebased on https://github.com/facebook/infer/commit/9348012dc324652685ea6a4196bb9e5f7ddc1d12.

I'll add some replies later to individual comments, but here are some points that are more important:

1. The test GreedyMeet Java has #804 (which illustrates precision) and another example that illustrates worst case. It's not that easy to hit the worst case, but it was indeed the case that the timeout [Config.max_meet] needed to be kicked down a bit to ensure that Infer still works fast. The lowered timeout is hit in <0.01% cases now.

2. I don't know if the prefix cache helps with overall times, but it has a high hit rate: >89% for 95% of the calls to `maximal_combine`. So, I think adding it was a good idea.

3. I'm not convinced that doing the vertex cover is worth it. The current algorithm covers all vertices in >99.85% cases. Let me know if you still want it, or if you want to discuss it.

The numbers from above are based on the attached log, which comes from running on Tomcat 9.0.0.M22.
[tomcat.log.txt](https://github.com/facebook/infer/files/1733784/tomcat.log.txt)

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
 
+module MaxCliqueState = Caml.Set.Make (struct
+  type t = int list (* invariant: increasing *)
+  let rec compare xs ys = match xs, ys with
```
done. thanks

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    for i = 0 to n - 1 do
+      for j = i + 1 to n - 1 do
+        (try
+          let z = plus_exc (by_idx i) (by_idx j) in
```
i changed this part to avoid effects (refs & exceptions)

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+  let n = List.length xs in
+  let by_idx =
+    let xs = Array.of_list xs in
+    Array.get xs in
```
doesn't exist anymore

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    | None -> raise Undefined
+    | Some z -> z in
+  let pairs = (* first, build graph with pairwise consistency *)
+    let h = Hashtbl.create (n * n) in
```
i turned this into the cache for prefixes, so now the value is used

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+  let compatible x xs = List.for_all xs ~f:(edge x) in
+  let first = filter_to n compatible [] 0 in
+  let module S = MaxCliqueState in
+  (S.add first S.empty, n, edge)
```
done, thanks

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    if !maxclique_count >= timeout then raise Timeout;
+    let xs, ng = next_maxclique g in (* xs is a pairwise consistent set *)
+    let zs = List.map ~f:by_idx xs in
+    (try result :=  (List.fold ~init ~f:plus_exc zs) :: !result
```
yes, too bad :p
but that would be a hack, and a nicer solution would be doing a vertex cover to which you allude later

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    let xs, ng = next_maxclique g in (* xs is a pairwise consistent set *)
+    let zs = List.map ~f:by_idx xs in
+    (try result :=  (List.fold ~init ~f:plus_exc zs) :: !result
+    with Undefined -> incr lost);
```
so, i'd say it's sound but incomplete.
With the modified version, there's a cache that would contain {A,B}, {B,C}, {A,C}, {B,C}, {C,D}, {B,D} mapped to their sums. You'd be able to use these to do a vertex cover. (In general, the cache would have all the prefixes that worked, all the singletons, and all pairs. Not entirely satisfying ...)

Except that, as I said in the main comment, this just doesn't seem to happen in practice. So, I'm not sure it's a good idea to make this code even more complicated. But, I'm willing to discuss if you disagree.

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    filter_to n ok_y [] 0 in
+  let do_y queue y =
+    let xy_compat x = x < y && edge x y in
+    let zs = List.filter xs ~f:xy_compat @ [y] in
```
meh ... ok
(maxclique doesn't need to decrement, though; unless I'm confused)

rgrig replied:
oh dear, i was confused.
given it's weekend, i suppose there won't be any race condition, but i'm still sorry for jumping the gun here

will fix soon

rgrig replied:
fixed now

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    let xy_compat x = x < y && edge x y in
+    let zs = List.filter xs ~f:xy_compat @ [y] in
+    let maxclique zs = (* is zs a maxclique for the subgraph induced by 0..y? *)
+      let rec loop u vs = u > y || (match u, vs with
```
right. that was quite silly

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+      let compatible z zs = List.for_all zs ~f:(edge z) in
+      filter_to n compatible (List.rev zs) (y + 1) in
+    if maxclique zs then S.add (fillup zs) queue else queue in
+  let queue = List.fold ~init:queue ~f:do_y ys in
```
I'm not sure I get this:
They _could_, but it seems to me it would be quite ugly. These two loops appear separated even in the pseudocode from the JYP paper.

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
   in
   res
 
+module MaxCliqueState = Caml.Set.Make (struct
```
done

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+  let compatible x xs = List.for_all xs ~f:(edge x) in
+  let first = filter_to n compatible [] 0 in
+  let module S = MaxCliqueState in
+  (S.add first S.empty, n, edge)
```
Yes, but it's nicer if the user doesn't need to know what the state is.
I added a `fold_stream` to make the idea explicit.

rgrig commented on infer/src/backend/dom.ml:
```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+      done
+    done; h in
+  let edge i j = Hashtbl.mem pairs (i, j) in
+  let maxclique_count = ref 0 in
```
removed

rgrig commented:
Wait a minute. I think I got confused with git while rebasing & handling conflilcts: code is OK, but the test is not the one that actually exercises worst case. Will fix.

jvillard commented:
Closing this because it's quite old. Please re-open if we still want to merge this.
